File: Data Structures & Algorithms/partition-equal-subset-sum/submission-0.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        totalSum = sum(nums)
        if totalSum % 2 == 1:
            return False
        
        def dfs(i, cur_sum, target):
            if i >= len(nums):
                return False
            
            if (cur_sum == target):
                return True

            return dfs(i + 1, cur_sum + nums[i], target) or dfs(i + 1, cur_sum, target)
        
        return dfs(0, 0, totalSum // 2)

File: Data Structures & Algorithms/combination-target-sum/submission-0.java
class Solution {
    public void dfs(int[] nums, int target, int i, List<Integer> subset, List<List<Integer>> subsets) {
        if (target == 0) {
            subsets.add(new ArrayList(subset));
            return;
        }

        if (target < 0 || i == nums.length) {
            return;
        }

        subset.add(nums[i]);
        dfs(nums, target - nums[i], i, subset, subsets);
        subset.remove(subset.size() - 1);
        dfs(nums, target, i + 1, subset, subsets);
    }

    public List<List<Integer>> combinationSum(int[] nums, int target) {
        List<List<Integer>> subsets = new ArrayList();
        dfs(nums, target, 0, new ArrayList(), subsets);
        return subsets;
    }
}
